test/hotspot/jtreg/runtime/os/TestTransparentHugePageUsage.java

```diff
@@ -0,0 +1,107 @@
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Optional;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import jdk.test.lib.process.ProcessTools;
+
+public class TestTransparentHugePageUsage {
+  private static final String[] fixedCmdLine = {
+    "-XX:+UseTransparentHugePages", "-XX:+AlwaysPreTouch",
+    "-Xlog:startuptime,pagesize,gc+heap=debug",
+    "-XX:+UseParallelGC", "-Xms1G", "-Xmx1G",
+  };
+
+  public static void main(String[] args) throws Exception {
+    ArrayList<String> cmdLine = new ArrayList<>(Arrays.asList(fixedCmdLine));
+    if (args.length > 0) {
+      cmdLine.add("-XX:ParallelGCThreads=" + args[0]);
+    }
+    cmdLine.add("runtime.os.TestTransparentHugePageUsage$CatSmaps");
+    ProcessBuilder builder = ProcessTools.createTestJavaProcessBuilder(cmdLine);
+    checkUsage(new BufferedReader(new InputStreamReader(builder.start().getInputStream())));
+  }
+
+  private static void checkUsage(BufferedReader reader) throws Exception {
+    final Pattern useThp = Pattern.compile(".*\\[info\\]\\[pagesize\\].+UseTransparentHugePages=1.*");
+    // Ensure THP is not disabled by OS.
+    if (reader.lines().filter(line -> useThp.matcher(line).matches()).findFirst().isPresent()) {
+      final Pattern heapAddr = Pattern.compile(".*\\sHeap:\\s.+base=0x0*(\\p{XDigit}+).*");
+      final Optional<Long> addr = reader.lines()
+          .map(line -> new SimpleEntry<String, Matcher>(line, heapAddr.matcher(line)))
+          .filter(e -> e.getValue().matches())
+          .findFirst()
+          .map(e -> Long.valueOf(e.getKey().substring(e.getValue().start(1), e.getValue().end(1)), 16));
+      if (!addr.isPresent()) throw new RuntimeException("Heap base was not found in smaps.");
+      // Match the start of a mapping, for example:
+      // 200000000-800000000 rw-p 00000000 00:00 0
+      final Pattern mapping = Pattern.compile("^(\\p{XDigit}+)-\\p{XDigit}+.*");
+      reader.lines()
+            .filter(line -> {
+                  Matcher matcher = mapping.matcher(line);
+                  if (matcher.matches()) {
+                    Long mappingAddr = Long.valueOf(line.substring(matcher.start(1), matcher.end(1)), 16);
+                    if (mappingAddr.equals(addr.get())) return true;
+                  }
+                  return false;
+                })
+            .findFirst();
+      final Pattern thpUsage = Pattern.compile("^AnonHugePages:\\s+(\\d+)\\skB");
+      final Optional<Long> usage = reader.lines()
+          .map(line -> new SimpleEntry<String, Matcher>(line, thpUsage.matcher(line)))
+          .filter(e -> e.getValue().matches())
+          .findFirst()
+          .map(e -> Long.valueOf(e.getKey().substring(e.getValue().start(1), e.getValue().end(1))));
+      if (!usage.isPresent()) throw new RuntimeException("The usage of THP was not found.");
+      if (usage.get() == 0) throw new RuntimeException("The usage of THP should not be zero.");
```

Comment:
The effect we would see without your patch would be small pages that are then converted to huge pages by khugepaged in its own time, right? So, maybe test that AnonHugePages == Size for the heap VMA?

Reply:
Unluckily, the size of AnonHugePages is still smaller than the heap size.